packages/compass-user-data/src/user-data.ts

```diff
@@ -2,39 +2,50 @@ import { promises as fs } from 'fs';
 import path from 'path';
 import { createLoggerAndTelemetry } from '@mongodb-js/compass-logging';
 import { getStoragePath } from '@mongodb-js/compass-utils';
+import type { z } from 'zod';
 
 const { log, mongoLogId } = createLoggerAndTelemetry('COMPASS-USER-STORAGE');
 
-type SerializeContent<T> = (content: T) => string;
-type DeserializeContent<T> = (content: string) => T;
+type SerializeContent<I> = (content: I) => string;
+type DeserializeContent = (content: string) => unknown;
+type GetFileName = (id: string) => string;
 
-type UserDataOptions<T = unknown> = {
+type UserDataOptions<I> = {
   subdir: string;
   basePath?: string;
-  serialize?: SerializeContent<T>;
-  deserialize?: DeserializeContent<T>;
+  serialize?: SerializeContent<I>;
+  deserialize?: DeserializeContent;
+  getFileName?: GetFileName;
 };
 
 type ReadOptions = {
   ignoreErrors: boolean;
 };
 
-export class UserData<T> {
+export class UserData<T extends z.Schema> {
   private readonly subdir: string;
   private readonly basePath?: string;
   private readonly serialize: SerializeContent<T>;
-  private readonly deserialize: DeserializeContent<T>;
-
-  constructor({
-    subdir,
-    basePath,
-    serialize = (content: T) => JSON.stringify(content, null, 2),
-    deserialize = JSON.parse,
-  }: UserDataOptions<T>) {
+  private readonly deserialize: DeserializeContent;
+  private readonly validator: z.ZodType<T>;
+  private readonly getFileName: GetFileName;
+
+  constructor(
+    getSchemaValidator: () => T,
```

Comment:
Again, only curious, no request for changes – is there a reason not to accept a schema validator directly here, and instead using a callback? I most commonly see this pattern when we want to call the callback at a later point or lazily, but I guess that doesn’t apply here?

Reply:
no specific reason why i designed it such. it does make it easy to directly accept the schema tbh. 
i can change it to that if that's fine with you.

Reply:
Yeah, no super strong preferences, just something that stuck out to me because it felt like a level of indirection without an obvious purpose

Reply:
clean up [e883aab](https://github.com/mongodb-js/compass/pull/4777/commits/e883aaba795a6430fb8e886c0b26da372ca5acb6)